client/src/main/java/io/pravega/client/connection/impl/AppendBatchSizeTrackerImpl.java

```diff
@@ -73,15 +79,28 @@ public int getAppendBlockSize() {
         if (numInflight <= 1) {
             return 0;
         }
-        double appendsInMaxBatch = Math.max(1.0, MAX_BATCH_TIME_MILLIS / millisBetweenAppends.getCurrentValue());
-        double targetAppendsOutstanding = MathHelpers.minMax(appendsOutstanding.getCurrentValue() * 0.5, 1.0,
-                                                             appendsInMaxBatch);
-        return (int) MathHelpers.minMax((long) (targetAppendsOutstanding * eventSize.getCurrentValue()), 0,
-                                        MAX_BATCH_SIZE);
+        double nanosPerAppend = nanosBetweenAppends.getCurrentValue();
+        double appendsInMaxBatchTime = (MAX_BATCH_TIME_MILLIS * NANOS_PER_MILLI) / nanosPerAppend;
+        double appendsInTime = Math.max(1.0, BASE_TIME_NANOS / nanosPerAppend);
+        double appendsInBatch = MathHelpers.minMax(appendsOutstanding.getCurrentValue() * OUTSTANDING_FRACTION, appendsInTime, appendsInMaxBatchTime);
+        int size = (int) (appendsInBatch * eventSize.getCurrentValue()) + BASE_SIZE;
+        return MathHelpers.minMax(size, 0, MAX_BATCH_SIZE);
     }
 
     @Override
     public int getBatchTimeout() {
         return MAX_BATCH_TIME_MILLIS;
     }
+    
+    private static int initializeConstant(String name, int defaultValue) {
+        String val = System.getenv(name);
+        if (val != null) {
+            try {
+                return Integer.parseInt(val);
+            } catch (NumberFormatException e) {
+                //Use default
```

Comment:
IMHO I would throw an exception, otherwise the user has no way to check that the configured value is in use.
I think that not everyone checks the log, you usually check logs only when you have a problem.

Reply:
Done